mdn/yari
# Generate error pages, e.g. 404

Fixes #1531

To test, run:
```
▶ yarn prepare-build
▶ yarn start
```

You can now go to a page like http://localhost:5000/en-US/docs/xxkpjadefd and you should see it in action. 
Also, look what got created in `client/build/en-us`.

Basically, the command `yarn build` now does 2 things:

1. Build all the "SPA" (aka. skeleton) pages and puts them into `client/build/en-us/_spas/`. (Note at the moment, there's only the 404.html but there'll probably others like home page and site-search)
2. All documents (according to `$BUILD_FOLDERSEARCH` or `$BUILD_FILES`)

You can use `yarn build --spas-only` or `yarn build --no-spas` to control that so it's not always doing those both two things. 

If you use http://localhost:3000/en-US/docs/xxkpjadefd you'd get what a writer would see. The `<Document/>` page starts to render but can't find the relevant `./index.json` so you get a "Loading Error" page. This isn't particularly important since it's rare and never applicable to production. 

- [x] Write end-to-end tests
- [x] Write docs explaining how all of this works (e.g. `docs/spas.md`)
- [x] Make the error page itself prettier
- [x] Document about the CLIish too `yarn build ...`

Also, once this lands you can now do
```
yarn build /path/to/file1/index.html /path/to/file2/index.html ...
```
to build specific files. That's so neat it ought to be documented and we can use this in the `pr-build.yml` on `mdn/content`.

The other major thing is that once this lands, we can train CloudFront to have a custom error behavior for the `*/docs*` Behavior. If a key doesn't exist on S3, it should fall back to reading `/en-us/_spas/404.html` from S3 and that needs to be quite carefully tested.
